File: app/src/at-linhle/java/com/asiantech/intern20summer1/intro/RainsOfReason.java
package com.asiantech.intern20summer1.intro;

public class RainsOfReason {
    /**
     * Given an array of integers, replace all the occurrences of elemToReplace with
     * substitutionElem.
     */
    int[] arrayReplace(int[] inputArray, int elemToReplace, int substitutionElem) {
        for (int i = 0; i < inputArray.length; i++) {
            if (inputArray[i] == elemToReplace) {
                inputArray[i] = substitutionElem;
            }
        }
        return inputArray;
    }

    /**
     * Check if all digits of the given integer are even.
     */
    boolean evenDigitsOnly(int n) {
        while (n != 0) {
            if (n % 10 % 2 != 0) {
                return false;
            }
            n = n / 10;
        }
        return true;
    }

    /**
     * Correct variable names consist only of English letters, digits and underscores and they
     * can't start with a digit.
     * Check if the given string is a correct variable name.
     */
    boolean variableName(String name) {
        for (int i = 0; i < name.length(); i++) {
            if (!('a' <= name.charAt(i) && name.charAt(i) <= 'z' ||
                    'A' <= name.charAt(i) && name.charAt(i) <= 'Z' ||
                    '0' <= name.charAt(i) && name.charAt(i) <= '9' ||
                    name.charAt(i) == '_')) {
                return false;
            }
        }
        return '0' > name.charAt(0) || name.charAt(0) > '9';
    }

    /**
     * Given a string, your task is to replace each of its characters by the next one in the
     * English alphabet; i.e. replace a with b, replace b with c, etc (z would be replaced by a).
     */
    String alphabeticShift(String inputString) {
        StringBuilder temp = new StringBuilder();
        for (int i = 0; i < inputString.length(); i++) {
            if (inputString.charAt(i) == 'z') {
                temp.append('a');
            } else {
                temp.append((char) ((int) inputString.charAt(i) + 1));
            }
        }
        return temp.toString();
    }

    /**
     * Given two cells on the standard chess board, determine whether they have the same color
     * or not.
     */
    boolean chessBoardCellColor(String cell1, String cell2) {
        char column1 = cell1.charAt(0), column2 = cell2.charAt(0);
        char row1 = cell1.charAt(1), row2 = cell2.charAt(1);
        return (column1 + row1) % 2 == (column2 + row2) % 2;
    }
}

File: app/src/at-vuongphan/java/kotlin/RainsofReason.kt
object RainsofReason {
    const val ZERO = 0
    const val TOW = 2
    const val Z = 'Z'
    const val CHARACTERA = 'a'
    const val CHARACTERZ = 'z'
    const val AUPPER = 'A'
    @JvmStatic
    fun main(args: Array<String>) {
        /**
         * Given an array of integers, replace all the occurrences of elemToReplace with substitutionElem.
         */
        println("25.Result:${arrayReplace(mutableListOf(1,2,1),1,3)}")
        /**
         * Check if all digits of the given integer are even
         */
        println("26.Result:${evenDigitsOnly(248622)}")
        /**
         * Correct variable names consist only of English letters, digits and underscores and they can't start with a digit.
         * Check if the given string is a correct variable name.
         */
        println("27.Result:${variableName("qq-q")}")
        /**
         *Given a string, your task is to replace each of its characters by the next one in the
         * English alphabet; i.e. replace a with b, replace b with c, etc (z would be replaced by a).
         */
        println("28.Result:${alphabeticShift("crazy")}")
        /**
         * Given two cells on the standard chess board, determine whether they have the same color or not.
         */
        println("29.Result:${chessBoardCellColor("A1","C3")}")
    }
    private fun chessBoardCellColor(cell1: String, cell2: String)
            = (cell1[0].toInt()+cell1[1].toInt()) % TOW == (cell2[0].toInt()+cell2[1].toInt()) % TOW
    private fun alphabeticShift(inputString: String): String {
        var string : String =""
        for(i in 0 until inputString.length){
            if(inputString.get(i) == CHARACTERZ) {
                string += CHARACTERA
            } else if(inputString.get(i) == Z){
                string += AUPPER
            }else{
                string = string +(inputString.get(i)+1).toChar()
            }
        }
        return string
    }
    private fun variableName(name: String): Boolean {
        val regex = """[a-zA-Z_][0-9a-zA-Z_]*${'$'}""".toRegex()
        if (name.matches(regex)) {
            return true;
        }
        return false;
    }
    private fun evenDigitsOnly(n: Int): Boolean {
        var string = n .toString()
        for(i in 0 until string.length){
            if(!(string[i].toInt() % TOW == ZERO)){
                return false
            }
        }
        return true
    }
    private fun arrayReplace(inputArray: MutableList<Int>, elemToReplace: Int, substitutionElem: Int): MutableList<Int> {
        for(i in ZERO until inputArray.size){
            if(inputArray[i] == elemToReplace){
                inputArray[i] = substitutionElem
            }
        }
        return inputArray
    }
}
